node/src/components/console.rs

```diff
@@ -0,0 +1,163 @@
+//! Unix console component.
+//!
+//! The console listens on a configurable unix socket for incoming connections and allows deep debug
+//! access to a running node via special commands.
+
+use std::{
+    fmt::{self, Display, Formatter},
+    fs, io,
+    path::{Path, PathBuf},
+};
+
+use datasize::DataSize;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tokio::{net::UnixListener, sync::watch};
+use tracing::debug;
+
+use super::Component;
+use crate::{
+    effect::{console::DumpConsensusStateRequest, EffectBuilder, EffectExt, Effects},
+    reactor::EventQueueHandle,
+    types::NodeRng,
+    utils::umask,
+    WithDir,
+};
+use util::ShowUnixAddr;
+
+mod command;
+mod tasks;
+mod util;
+
+/// Unix console component.
+#[derive(Debug, DataSize)]
+pub(crate) struct Console {
+    /// Sender, when dropped, will cause server and client connections to exit.
+    #[data_size(skip)]
+    #[allow(dead_code)] // only used for its `Drop` impl.
+    shutdown_sender: watch::Sender<()>,
+}
+
+/// Unix console configuration.
+#[derive(Clone, DataSize, Debug, Deserialize)]
+pub(crate) struct Config {
+    /// Whether or not the console is enabled.
+    enabled: bool,
+    /// Path to listen on.
+    socket_path: PathBuf,
+    /// `umask` to apply before creating the socket.
+    socket_umask: u16,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            enabled: false,
+            socket_path: "debug.socket".into(),
+            socket_umask: 0o077,
```

Comment:
How is that `u16`? 😅 I haven't seen this format before. What does the `o` indicate to the compiler?

Reply:
Octal notation, the `0o` is to octal what `0x` is to hexadecimal. Other languages use `0077` == `077` (leading zero indicates octal), which is confusing. Rust thus requires the explicit `o`.

Reply:
Thanks! I didn't know that :)